### PR TITLE
[`Mixtral` & `Mistral`] Add support for sdpa

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?

Adds the SDPA attention for both classes cc @younesbelkada for visibility 😉 Will help for fast LLava

### CODE REVIEW

File: src/transformers/models/mistral/modeling_mistral.py [hunk 1]
```diff
@@ -891,18 +981,27 @@ def forward(
...
             inputs_embeds = self.embed_tokens(input_ids)
 
-        if attention_mask is not None and self._use_flash_attention_2 and use_cache:
+        if attention_mask is not None and self._attn_implementation == "flash_attention_2" and use_cache:
             is_padding_right = attention_mask[:, -1].sum().item() != batch_size
             if is_padding_right:
                 raise ValueError(
                     "You are attempting to perform batched generation with padding_side='right'"
-                    " this may lead to unexpected behaviour for Flash Attention version of Mistral. Make sure to "
+                    " this may lead to unexpected behaviour for Flash Attention version of Mixtral. Make sure to "
```
Comment (younesbelkada) [inline src/transformers/models/mistral/modeling_mistral.py:989]:
```suggestion
                    " this may lead to unexpected behaviour for Flash Attention version of Mistral. Make sure to "
```

File: src/transformers/models/mistral/modeling_mistral.py [hunk 2]
```diff
@@ -820,7 +910,7 @@ def __init__(self, config: MistralConfig):
         self.layers = nn.ModuleList(
             [MistralDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
         )
-        self._use_flash_attention_2 = config._attn_implementation == "flash_attention_2"
```
Comment (younesbelkada) [inline src/transformers/models/mistral/modeling_mistral.py:823]: Any reason this has been removed? Shall we keep it for consistency with llama: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L961 ? I found it much more readable to have `_use_flash_attention_2` instead of checking `self._attn_implementation == "flash_attention_2"` each time

**Review by younesbelkada — APPROVED** [review]
Thanks ! 
I don"t see why sliding window attention shouldn't be supported with SDPA because the only difference vs the eager attention implementation is on the attention mask. Passing arbitrary attention masks in SDPA should be supported without any problem IMO

File: src/transformers/models/mistral/modeling_mistral.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/mistral/modeling_mistral.py:823]: No because we need to have one for sdpa as well which does not make sense

**Comment by ehuaa** [comment]
> Thanks ! I don"t see why sliding window attention shouldn't be supported with SDPA because the only difference vs the eager attention implementation is on the attention mask. Passing arbitrary attention masks in SDPA should be supported without any problem IMO

I have the same problem here, why sdpa not support window attention? Is there any problems not been solved? @ArthurZucker

**Comment by younesbelkada** [comment]
@ehuaa the way the window attention is implemented in Mistral original code base is by changing the attention mask to a "more custom" attention mask to not attend to tokens that are before `sliding_windows`. Check out more by looking into the details of this method: https://github.com/huggingface/transformers/blob/d90acc16437e8c9e45e068fa1cc1a263b9a7208f/src/transformers/modeling_attn_mask_utils.py#L145
The point that I tried to convey is that passing that attention mask is supported I think in SDPA so you can implicitly get SDPA + sliding window attention by just passing that correct attention mask. Let me know if this makes sense to you!

**Comment by ehuaa** [comment]
> @ehuaa the way the window attention is implemented in Mistral original code base is by changing the attention mask to a "more custom" attention mask to not attend to tokens that are before `sliding_windows`. Check out more by looking into the details of this method:
> 
> https://github.com/huggingface/transformers/blob/d90acc16437e8c9e45e068fa1cc1a263b9a7208f/src/transformers/modeling_attn_mask_utils.py#L145
> 
> 
> The point that I tried to convey is that passing that attention mask is supported I think in SDPA so you can implicitly get SDPA + sliding window attention by just passing that correct attention mask. Let me know if this makes sense to you!

@younesbelkada Thank you for your quick reply! Your solution above can pass a custom mask to sdpa, and i think this way is the same as passing sliding_window param to this function.
https://github.com/huggingface/transformers/blob/main/src/transformers/models/mistral/modeling_mistral.py#L1006-L1023
image
